drivers/sensor/nordic/qdec_nrfx/qdec_nrfx.c

```diff
@@ -173,81 +174,78 @@ static const struct sensor_driver_api qdec_nrfx_driver_api = {
 	.trigger_set  = qdec_nrfx_trigger_set,
 };
 
-#ifdef CONFIG_PM_DEVICE
-static int qdec_nrfx_pm_action(const struct device *dev,
-			       enum pm_device_action action)
+static int qdec_pm_suspend(const struct device *dev)
 {
 	const struct qdec_nrfx_config *config = dev->config;
-	int ret = 0;
+
+	nrfx_qdec_disable(&config->qdec);
+	qdec_nrfx_gpio_ctrl(dev, false);
+
+	if (!IS_ENABLED(CONFIG_PM_DEVICE)) {
+		return 0;
+	}
+
+	return pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
+}
+
+static int qdec_pm_resume(const struct device *dev)
+{
+	const struct qdec_nrfx_config *config = dev->config;
+	int ret;
+
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	qdec_nrfx_gpio_ctrl(dev, true);
+	nrfx_qdec_enable(&config->qdec);
+
+	return 0;
+}
+
+static int qdec_nrfx_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	int ret = -ENOTSUP;
 
 	switch (action) {
 	case PM_DEVICE_ACTION_RESUME:
-		ret = pinctrl_apply_state(config->pcfg,
-					  PINCTRL_STATE_DEFAULT);
-		if (ret < 0) {
-			return ret;
-		}
-		qdec_nrfx_gpio_ctrl(dev, true);
-		nrfx_qdec_enable(&config->qdec);
-		break;
-
-	case PM_DEVICE_ACTION_TURN_OFF:
-		/* device must be uninitialized */
-		nrfx_qdec_uninit(&config->qdec);
-		ret = pinctrl_apply_state(config->pcfg,
-					  PINCTRL_STATE_SLEEP);
-		if (ret < 0) {
-			return ret;
-		}
+		ret = qdec_pm_resume(dev);
 		break;
 
 	case PM_DEVICE_ACTION_SUSPEND:
-		/* device must be suspended */
-		nrfx_qdec_disable(&config->qdec);
-		qdec_nrfx_gpio_ctrl(dev, false);
-		ret = pinctrl_apply_state(config->pcfg,
-					  PINCTRL_STATE_SLEEP);
-		if (ret < 0) {
-			return ret;
+		if (IS_ENABLED(CONFIG_PM_DEVICE)) {
+			ret = qdec_pm_suspend(dev);
 		}
 		break;
 	default:
-		return -ENOTSUP;
+		break;
 	}
 
 	return ret;
 }
-#endif /* CONFIG_PM_DEVICE */
 
 static int qdec_nrfx_init(const struct device *dev)
 {
-	const struct qdec_nrfx_config *dev_config = dev->config;
-
-	dev_config->irq_connect();
+	const struct qdec_nrfx_config *config = dev->config;
+	int ret;
+	nrfx_err_t nerr;
 
-	int err = pinctrl_apply_state(dev_config->pcfg, PINCTRL_STATE_DEFAULT);
+	config->irq_connect();
 
-	if (err < 0) {
-		return err;
+	nerr = nrfx_qdec_init(&config->qdec, &config->config, qdec_nrfx_event_handler, (void *)dev);
+	if (nerr != NRFX_SUCCESS) {
+		return (nerr == NRFX_ERROR_INVALID_STATE) ? -EBUSY : -EFAULT;
 	}
 
-	nrfx_err_t nerr = nrfx_qdec_init(&dev_config->qdec,
-					 &dev_config->config,
-					 qdec_nrfx_event_handler,
-					 (void *)dev);
-
-	if (nerr == NRFX_ERROR_INVALID_STATE) {
-		LOG_ERR("qdec already in use");
-		return -EBUSY;
-	} else if (nerr != NRFX_SUCCESS) {
-		LOG_ERR("failed to initialize qdec");
-		return -EFAULT;
+	if (IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME)) {
+		ret = qdec_pm_suspend(dev);
+		if (ret < 0) {
+			return ret;
+		}
```

Comment:
note: it would be nice if nrfx API just initialized device state, not performing any enablement. This would simplify driver design and avoid unnecessary transitions.

Reply:
nrfx after init ends up in off state but gpio part needs to be configured. I reworked it to not call suspend in init but added `qdec_nrfx_gpio_ctrl(dev, false);` and pinctrl setting to init function.